src/runtime/storageNG/tests/store-sequence-test.ts

```diff
@@ -0,0 +1,170 @@
+    return new MockDriver<Data>(storageKey, exists);
+  }
+}
+
+describe('Store Flow', async () => {
+  // Tests a model resync request happening synchronously with model updates from the driver
+  it('services a model request and applies 2 models', async () => {
+    const sequenceTest = new SequenceTest<ActiveStore<CRDTCountTypeRecord>>();
+    sequenceTest.setTestConstructor(() => {
+      DriverFactory.clearRegistrationsForTesting();
+      DriverFactory.register(new MockStorageDriverProvider());
+    
+      const store = new Store('string', Exists.ShouldCreate, null, StorageMode.Direct, CRDTCount);
+      const activeStore = store.activate();
+      return activeStore;
+    });    
+
+    const onProxyMessage = sequenceTest.registerInput('onProxyMessage', 1, 
+      {type: ExpectedResponse.Constant, response: true});
+    const onReceive = sequenceTest.registerInput('onReceive', 1, {type: ExpectedResponse.Void}); 
+
+    const send = sequenceTest.registerOutput('driver.send', 
+      {
+        type: ExpectedResponse.Defer,
+        default: true,
+      }, SequenceOutput.Replace);
+      
+    const idVar = sequenceTest.registerVariable(-1);
+    const isSyncRequest = sequenceTest.registerVariable(false);
+    const model = sequenceTest.registerVariable(() => new CRDTCount(), true);
+  
+    const on = sequenceTest.registerOutput('on',
+      {
+        type: ExpectedResponse.Constant,
+        response: true,
+        onOutput: (value: ProxyMessage<CRDTCountTypeRecord>) => {
+          const syncRequest = sequenceTest.getVariable(isSyncRequest);
+          if (syncRequest) {
+            assert.equal(value.type, ProxyMessageType.ModelUpdate);
+            sequenceTest.setVariable(isSyncRequest, false);
+          } else {
+            if (value.type === ProxyMessageType.Operations) {
+              const currModel: CRDTCount = sequenceTest.getVariable(model);
+              for (const operation of value.operations) {
+                assert.isTrue(currModel.applyOperation(operation));
+              }
+            } else {
+              assert.fail();
+            }
+          }
+        }
+      }, SequenceOutput.Register, idVar);
+      
+    const storageProxyChanges: {inputFn: () => ProxyMessage<CRDTCountTypeRecord>, variable: {}}[] 
+      = [{inputFn: () => ({type: ProxyMessageType.SyncRequest, id: sequenceTest.getVariable(idVar)}), 
+          variable: {[isSyncRequest]: true}}]; 
+
+    const makeModel = (meCount: number, themCount: number, meVersion: number, themVersion: number): CountData => 
+        ({values: new Map([['me', meCount], ['them', themCount]]), 
+          version: new Map([['me', meVersion], ['them', themVersion]])});
+    const driverChanges = [
+      {output: {[send]: false}},
+      {input: makeModel(7, 12, 3, 4), output: {[send]: true}},
+      {output: {[send]: false}},
+      {input: makeModel(8, 12, 4, 4), output: {[send]: true}}
+    ];
```

Comment:
I presume the ordering is significant here... why are we expecting outputs before inputs?

Reply:
This says: "for the object under test's onReceive method and the send method on its 'driver' property:
(1) force the send method's return value to be 'false' for all calls to it
((1a)) introduce a potential async boundary
(2) send a model to onReceive; also force the sne method's return value to be 'true' for all calls to it
((2a)) introduce a potential async boundary
etc."